### PR TITLE
CorfuTable Micro-benchmark

### PR DESCRIPTION
**Pull request by Maithem**
A microbenchmark for CorfuTable written in JMH.

### CODE REVIEW

File: benchmark/src/main/java/org/corfudb/benchmark/Common.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+
+import java.util.Map;
+
+@State(Scope.Benchmark)
+public class Common {
+    Map<Long, byte[]> map;
+    CorfuRuntime rt;
+    final BitsStreamGenerator RND = new MersenneTwister();
```
Comment (corfu-repo) [inline benchmark/src/main/java/org/corfudb/benchmark/Common.java]: MINOR Rename this field "RND" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00116)

File: benchmark/src/main/java/org/corfudb/benchmark/Common.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    @Setup(Level.Trial)
+    public void setup() {
+        rt = new CorfuRuntime("localhost:9000").connect();
+        map = rt.getObjectsView()
+                .build()
+                .setStreamName(Integer.toString(RND.nextInt()))
+                .setType(SMRMap.class)
```
Comment (corfu-repo) [inline benchmark/src/main/java/org/corfudb/benchmark/Common.java]: MINOR Remove this use of "SMRMap"; it is deprecated. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **e4f8b86**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #941 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/wgWKD3RxTXUE7pFvtxf0awKCgCvITRtD)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **3850380**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #941 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/2yCQ6RGuMrdLEUJU3WCIy8rHuwpgbNuT)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **7c93ddd**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #941 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/3oMKGwz0dyHsjitCXdFsj77cTPxnj6tW)

File: test/src/test/java/org/corfudb/runtime/collections/CorfuTableTest.java [hunk 3]
```diff
@@ -23,6 +23,7 @@
         @Getter
         final CorfuTable.IndexFunction<String, String, String> indexFunction;
 
+
```
Comment (no2chem) [inline test/src/test/java/org/corfudb/runtime/collections/CorfuTableTest.java]: Don't need this empty line

File: benchmark/src/main/java/org/corfudb/benchmark/Main.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+public class Main {
+    public static void main(String[] args) throws RunnerException {
+        Options opt = new OptionsBuilder()
```
Comment (no2chem) [inline benchmark/src/main/java/org/corfudb/benchmark/Main.java]: Let's use docopt to take some arguments from the command line and provide some documentation.

File: benchmark/src/main/java/org/corfudb/benchmark/CorfuTableBenchmark.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+            }
+
+            // Generate a random subset for the right keys. This is
+            // required to create collisions in the secondary index (i.e
+            // one key in the secondary index can reference multiple values)
+            Collections.shuffle(leftKeys);
+            rightKeys = leftKeys.subList(0, (int) (num * subsetRatio));
+        }
+
+        void write(int batch) {
```
Comment (corfu-repo) [inline benchmark/src/main/java/org/corfudb/benchmark/CorfuTableBenchmark.java]: MAJOR Remove this unused method parameter "batch". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1172)

File: benchmark/src/main/java/org/corfudb/benchmark/CorfuTableBenchmark.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
+package org.corfudb.benchmark;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.reflect.TypeToken;
+import lombok.Data;
+import lombok.Getter;
```
Comment (corfu-repo) [inline benchmark/src/main/java/org/corfudb/benchmark/CorfuTableBenchmark.java]: MINOR Remove this unused import 'lombok.Getter'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: benchmark/src/main/java/org/corfudb/benchmark/CorfuTableBenchmark.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
```
Comment (corfu-repo) [inline benchmark/src/main/java/org/corfudb/benchmark/CorfuTableBenchmark.java]: MINOR Remove this unused import 'lombok.RequiredArgsConstructor'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: benchmark/src/main/java/org/corfudb/benchmark/Common.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+import org.apache.commons.math3.random.MersenneTwister;
+import org.corfudb.runtime.CorfuRuntime;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.State;
+
+import static org.openjdk.jmh.annotations.Scope.Benchmark;
+
+@State(Benchmark)
+public class Common {
+    final BitsStreamGenerator RND = new MersenneTwister();
```
Comment (corfu-repo) [inline benchmark/src/main/java/org/corfudb/benchmark/Common.java]: MINOR Rename this field "RND" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00116)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **f43e1bc**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #941 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/7A4k6zVwMlP6Hqb885z1pZMQzF8vn2Eb)

**Comment by Maithem** [comment]
@no2chem please review again, I addressed all your comments.

File: benchmark/src/main/java/org/corfudb/benchmark/Common.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    CorfuRuntime rt;
+
+    @Param("1000")
+    int num;
+
+    @Param("9000")
+    String port;
+
+    public void setup() {
+        System.out.println("stats");
```
Comment (corfu-repo) [inline benchmark/src/main/java/org/corfudb/benchmark/Common.java]: MAJOR Replace this use of System.out or System.err by a logger. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS106)

File: benchmark/src/main/java/org/corfudb/benchmark/Common.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        System.out.println(port);
```
Comment (corfu-repo) [inline benchmark/src/main/java/org/corfudb/benchmark/Common.java]: MAJOR Replace this use of System.out or System.err by a logger. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS106)

File: benchmark/src/main/java/org/corfudb/benchmark/Common.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+        System.out.println(num);
```
Comment (corfu-repo) [inline benchmark/src/main/java/org/corfudb/benchmark/Common.java]: MAJOR Replace this use of System.out or System.err by a logger. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS106)

File: benchmark/src/main/java/org/corfudb/benchmark/Benchmark.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+public class Benchmark {
+    private static final String doc = "Benchmark.\n"
```
Comment (corfu-repo) [inline benchmark/src/main/java/org/corfudb/benchmark/Benchmark.java]: CRITICAL Rename this constant name to match the regular expression '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00115)

File: benchmark/src/main/java/org/corfudb/benchmark/Benchmark.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+import org.corfudb.util.GitRepositoryState;
+import org.docopt.Docopt;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+import java.util.HashMap;
```
Comment (corfu-repo) [inline benchmark/src/main/java/org/corfudb/benchmark/Benchmark.java]: MINOR Remove this unused import 'java.util.HashMap'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 5 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **30e0a1c**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #941 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/l4I6M2UXdOzfEGa7lbFrcBYMTUWVk71h)

**Comment by no2chem** [comment]
not sure why this autoclosed?

**Comment by no2chem** [comment]
If I did, it wasn't intentional.

I pushed a benchmark2 branch, but it definitely was not a force push

**Pull request closed without merging**